### ISSUE TITLE
Browser: worker not mocking requests when `vi.mock` present

### ISSUE THREAD

**Issue by charlieforward9** [kettanaito/msw#2398]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 18 or higher

### Browsers

Chromium (Chrome, Brave, etc.)

### Reproduction repository

https://github.com/Agriculture-Intelligence/msw-vitest-browser-issue

### Reproduction steps

* Install `aws-amplify`, `vitest`, `msw`
* Configure `vitest` and `msw` to support browser-based testing/mocking
* Try mocking a call to the `downloadData` command from `aws-amplify/storage` module
* Notice the non-detailed `NETWORK ERR`

### Current behavior

Functional in development and node/jsdom testing, when configuration is changed to browser-based, fails.

### Expected behavior

I expect the browser based mocking to work exactly the same as node based testing.

**Comment by charlieforward9** [comment]
Looking at more of my tests, it seems like the real problem is that my handlers are not properly intercepting/mocking the network calls. 

It appears to fire the request and match correctly, but not mock properly and end up somehow bypassing?

Am I configuring the `worker` incorrectly?

Please take a look at the sample repo so I can fix it and contribute it to an msw/example browser mode setup. Thank you

**Comment by charlieforward9** [comment]
I am also getting this at some places in my console when I run my tests... 
```
 ... Warning: intercepted a request without a matching request handler:

  • GET /node_modules/...
```

i did not have any issues with any of the mocking when I was using the `setupServer` config, but now with `setupWorker`, there are a bunch of miscellaneous issues relating to external calls, including calls to `http://localhost:63315/node_modules/`...

**Comment by bdpartridge** [comment]
I'm hitting a seemingly similar issue while trying to use msw with vitest browser mode. My request handlers are being called, but ultimately msw is bypassing all requests, so all my mocked requests are resulting in 404 responses.

**Comment by bdpartridge** [comment]
> I'm hitting a seemingly similar issue while trying to use msw with vitest browser mode. My request handlers are being called, but ultimately msw is bypassing all requests, so all my mocked requests are resulting in 404 responses.

I just realized that vitest browser is setting up its own msw request handler when `vi.mock(...)` is used to mock modules. Somehow this is interfering with my own request handlers, so it looks like this might be an issue with vitest browser and not msw.

**Comment by charlieforward9** [comment]
> vitest browser is setting up its own msw request handler when vi.mock(...) is used to mock modules. Somehow this is interfering with my own request handlers

@bdpartridge is it? Take a look at my [reproduction](https://github.com/Agriculture-Intelligence/msw-vitest-browser-issue). I don't use `vi.mock` and still get issues.

**Comment by bdpartridge** [comment]
@charlieforward9 You're right. It looks like yours is a different issue. I tried out your reproduction and noticed that the `beforeAll` hook in the `vitest.setup.ts` file wasn't awaiting the worker to start. After making this change:

```
beforeAll(async () => {
  Amplify.configure(amplifyconfig);
  await worker.start();
});
```

…I see the mocked 200 response for `POST https://cognito-identity.us-east-2.amazonaws.com/`, but there seems to be a separate CORS issue related to the preflight request to `https://mock-staging.s3.us-east-2.amazonaws.com/path/to/file.json`.

Anyway, sorry for adding noise to this thread. Good luck resolving this.

**Comment by charlieforward9** [comment]
@bdpartridge Yeah, I was under the impression that if we created the respective handler in msw, all network related operations would be skipped and handled by the msw handler. This is what is confusing me the most. Why am i getting network problems when implementing the mocks in msw?

**Comment by charlieforward9** [comment]
I mocked the preflight request with `http.options`, but it did not reveal anything novel. I still do not understand why it will work with the `setupServer` but not `setupWorker` in browser mode...

Could someone please point me to how to debug this better? I need to figure this out.

EDIT: I dont see `[MSW] Mocking enabled` log in my console, but I do see `[MSW] Mocking disabled` at the end. Would be great to get some eyes on this.

**Comment by charlieforward9** [comment]
@bdpartridge seems to be right

> I just realized that vitest browser is setting up its own msw request handler when vi.mock(...) is used to mock modules. Somehow this is interfering with my own request handlers, so it looks like this might be an issue with vitest browser and not msw.

The reason it wasnt working on my repro without `vi.mock` calls was because [I was originally returning `undefined` in the handler](https://github.com/Agriculture-Intelligence/msw-vitest-browser-issue/commit/5405584825aa65caab8a33719083a1199ebc0f7c#diff-bb09d52c957fd04a8ed1a9f64c0cd6da4bb6c207f92a0fa1bb81a0e957aeaf49L40:~:text=%2D-,return%20undefined%3B,-18), leading to it bypassing the call.

If you think this isnt a problem with msw, have you opened an issue with vitest yet? Any workarounds other than commenting out `vi.mock`?

**Comment by e7h4n** [comment]
I also encountered this issue, here is the minimal reproduction I tried: https://github.com/e7h4n/msw-vitest-failure/blob/main/index.test.ts

**Comment by louneskmt** [comment]
Same here, commenting out all the `vi.mock` makes the interceptors work again. Otherwise the handlers are correctly executed but for some reason the requests still end up a 404.

**Comment by vafanassieff** [comment]
If we use `vi.doMock` inside the test and import our components with `await import('file.tsx')` it does work.
Maybe hoisting `vi.mock` to the top-level break something?

**Comment by kettanaito** [comment]
Hi, everyone. Thanks for sharing your details and reproduction repos! Much appreciated.

This one is most likely on MSW's side, but it needs proper investigation before we can know for certain. Since Vitest also uses an MSW worker under the hood for `vi.mock()`, there's a slight chance that something goes off and request handlers conflict (we've covered a few of related issues in the past on Vitest's side; they're fixed now). 

## How you can help

1. In the reproduction scenario, log out `console.log(worker.handlersController.currentHandlers())`. What does it print you? You should see the list of _your_ handlers. 
2. Grab the handler you expect to match (e.g. by index: `currentHandlers()[0]`) and run `handler.run()`. Does it returned the mocked response you expect? If not, what does it return?

**Comment by vafanassieff** [comment]
1. 
Without a `vi.mock` in my tests
```
[
  HttpHandler {
    "__kind": "RequestHandler",
    "info": {
      "callFrame": "http://localhost:63315/@fs/Users/afa/code/afa/msw-trpc/packages/msw-trpc/src/handler.ts:36:14",
      "header": "GET /http:\/\/localhost:3000\/trpc\/userById$/",
      "method": "GET",
      "path": /http:\/\/localhost:3000\/trpc\/userById$/,
    },
    "isUsed": false,
    "options": {},
    "resolver": [Function anonymous],
    "resolverIterator": undefined,
    "resolverIteratorResult": undefined,
  },
]
```

With a `vi.mock`

```
[
  HttpHandler {
    "__kind": "RequestHandler",
    "info": {
      "callFrame": "http://localhost:63315/@fs/Users/afa/code/afa/msw-trpc/packages/msw-trpc/src/handler.ts:36:14",
      "header": "GET /http:\/\/localhost:3000\/trpc\/userById$/",
      "method": "GET",
      "path": /http:\/\/localhost:3000\/trpc\/userById$/,
    },
    "isUsed": false,
    "options": {},
    "resolver": [Function anonymous],
    "resolverIterator": undefined,
    "resolverIteratorResult": undefined,
  },
]
```

I have an error when executing the function

2.

```
TypeError: Cannot read properties of undefined (reading 'request')
    at HttpHandler.run (http://localhost:63315/@fs/Users/afa/code/afa/msw-trpc/node_modules/.pnpm/msw@2.7.1_@types+node@22.13.4_typescript@5.7.3/node_modules/msw/lib/core/handlers/RequestHandler.mjs?v=8f961d5b:77:63)
    at http://localhost:63315/Users/afa/code/afa/msw-trpc/packages/test-browser/src/app.test.tsx?import&browserv=1740212511749:49:21
    at async http://localhost:63315/@fs/Users/afa/code/afa/msw-trpc/node_modules/.pnpm/@vitest+runner@3.0.6/node_modules/@vitest/runner/dist/index.js?v=8f961d5b:558:22
```

**Comment by bdpartridge** [comment]
It looks like vitest runs its own worker setup when `vi.mock(...)` is called. Is it safe to call `workser.start(...)` when the worker has already been started? Even if it is, if the vitest request handler used for module mocking runs first, it creates a passthrough `Response` for any requests that haven't been mocked, which means other handlers would be skipped anyway.

Interestingly, the `@vitest/mocker` package that's used by vitest to mock modules [appears to support passing a pre-configured worker instance](https://github.com/vitest-dev/vitest/blob/1a8b4337e3e6981f358b119dddca2108ae08e180/packages/mocker/src/browser/interceptor-msw.ts#L31), but it doesn't look like there's a way to use that option from the API exposed by the main vitest package.

For now, I'm working around this by specifying aliases in the vitest config. That works okay for my use case since none of my module mocks are test-specific at the moment.

**Comment by moleske** [comment]
> For now, I'm working around this by specifying aliases in the vitest config. That works okay for my use case since none of my module mocks are test-specific at the moment.

@bdpartridge Could you provide an example of this?  We're hitting the same issue and are curious on this work around

**Comment by bdpartridge** [comment]
> Could you provide an example of this? We're hitting the same issue and are curious on this work around

@moleske We're [defining custom aliases](https://vitest.dev/config/#alias) in the vitest config to replace the modules we'd like to mock when running inside tests. I'm not sure how you've defined your vitest config, but assuming you've added it directly to your vite.config.js file, that would look something like this:

```js
// vite.config.js

import {resolve} from "node:path";

export default defineConfig({
  resolve: {
    alias: [
      {
        find: "$src",
        replacement: resolve(import.meta.dirname, "src"),
      },
    ],
  },
  test: {
    alias: [
      {
        find: "$src/my/module",
        replacement: resolve(import.meta.dirname, "test/mocks/my/module"),
      },
    ],
  },
});
```

This workaround is only useful if the module mock can be used in all your tests, although you could introduce additional workspaces to work around that limitation since each workspace can define its own aliases. Note that we're using vite 5.4.9 and vitest 3.0.4. Hopefully this helps.

**Comment by charlieforward9** [comment]
This has seemingly been fixed in https://github.com/vitest-dev/vitest/pull/7576 , release 3.1.0
